### PR TITLE
refactor: migrate ScreenLockConfigView to a function component

### PR DESCRIPTION
**Pull request by diegolmello**
## Proposed changes

Migrates the Screen Lock settings screen (`ScreenLockConfigView`) — which controls the auto-lock timeout and the biometry (Face ID / Touch ID / fingerprint) toggle — from a class component to a function component with React hooks. Part of the ongoing Migrate to Hooks effort. Behavior is preserved.

Notable points for review:

- The old class used `setState(updater, callback)` and read the just-updated state inside the callback to decide whether to persist, prompt for a passcode, or undo. The function version computes the next value explicitly and persists that value, so there is no stale-state read after a state setter — the classic hooks pitfall this migration had to avoid.
- Cancelling passcode setup now undoes the auto-lock toggle inline (turns it off and saves off) instead of recursively re-invoking the toggle, so the toggle analytics event fires once per user action instead of twice. End state is unchanged (auto-lock OFF and persisted OFF).
- Removed a dead `observable` field that was declared and unsubscribed but never assigned anywhere — no real subscription was dropped.
- Screen registration in `InsideStack` and `MasterDetailStack` was updated to the current static-config navigation pattern (matching the already-migrated `DisplayPrefsView`), replacing the static `navigationOptions` with a `useLayoutEffect` that sets the title.

## Issue(s)

https://rocketchat.atlassian.net/browse/NATIVE-27

## How to test or reproduce

Automated:

- `TZ=UTC pnpm test --testPathPattern='ScreenLockConfigView'` — 5 tests: render after init, auto-lock enable persists, passcode-cancel undo, biometry toggle, and auto-lock-time change.

Manual QA (Settings → Security and privacy → Screen lock):

- Toggle auto-lock on, set a passcode, then confirm it persists across an app restart.
- Cancel passcode setup and confirm auto-lock returns to OFF.
- Toggle biometry and confirm the preference sticks.

## Screenshots

No UI changes — this is a behavior-preserving refactor.

## Types of changes

- [x] Improvement (non-breaking change which improves a current function)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

The core correctness property to preserve in any future change here: side effects (DB save, biometry read) must act on the computed next value, not on state read right after a `useState` setter.

https://claude.ai/code/session_01PA87xg21JfSiVH6cZbeceC

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reliability of screen-lock settings: enabling auto-lock now verifies passcode availability, persists the correct auto-lock state and duration, and rolls back safely if enabling is canceled.
  * Kept biometry toggle behavior consistent with saved preferences and refreshed UI state.
  * Updated modal navigation so screen-lock options and admin-enforced durations apply consistently.
* **Refactor**
  * Converted the screen-lock configuration view to a hook-based implementation.
* **Tests**
  * Added/expanded coverage for rendering, auto-lock/duration updates, biometry toggling, and admin-forced locking behavior.

### CODE REVIEW

File: app/views/ScreenLockConfigView.tsx [hunk 1]
```diff
@@ -28,197 +29,157 @@ interface IItem {
...
+	const serverRecord = useRef<TServerModel | undefined>(undefined);
+
+	const [autoLock, setAutoLock] = useState(false);
+	const [autoLockTime, setAutoLockTime] = useState<number | null>(null);
+	const [biometry, setBiometry] = useState(DEFAULT_BIOMETRY);
+	const [biometryLabel, setBiometryLabel] = useState<string | null>(null);
+
+	useLayoutEffect(() => {
+		navigation.setOptions({ title: I18n.t('Screen_lock') });
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (diegolmello) [inline app/views/ScreenLockConfigView.tsx]: is it needed?

File: app/views/ScreenLockConfigView.tsx [hunk 2]
```diff
@@ -28,197 +29,157 @@ interface IItem {
...
+					serverRecord.current?.autoLockTime === null ? DEFAULT_AUTO_LOCK : serverRecord.current?.autoLockTime ?? null
+				);
+				setBiometry(userPreferences.getBool(BIOMETRY_ENABLED_KEY) ?? DEFAULT_BIOMETRY);
+			} catch {
+				// noop
+			}
+			setBiometryLabel(await supportedBiometryLabel());
+		};
+		init();
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (diegolmello) [inline app/views/ScreenLockConfigView.tsx]: is it needed?

File: app/views/ScreenLockConfigView.tsx [hunk 3]
```diff
@@ -252,48 +211,39 @@ class ScreenLockConfigView extends Component<IScreenLockConfigViewProps, IScreen
...
+						<>
+							<List.Separator />
+							<List.Item title='Local_authentication_change_passcode' onPress={handleChangePasscode} showActionIndicator />
+						</>
+					) : null}
+					<List.Separator />
+					<List.Info info='Local_authentication_info' />
+				</List.Section>
+				{renderBiometry()}
+				{renderAutoLockItems()}
```
Comment (diegolmello) [inline app/views/ScreenLockConfigView.tsx:226]: Evaluate if they should be moved to components

**Comment by Shevilll** [comment]
Hi cooked-pair,

Thanks for submitting this pull request! Migrating `ScreenLockConfigView` to a functional component using hooks and Redux toolkit selectors is a fantastic improvement for modernizing the codebase. 

Here are a few technical and constructive suggestions to further polish this migration:

### 1. Optimize Callback Stability with `useCallback`
Currently, functions like `toggleAutoLock`, `toggleBiometry`, `changeAutoLockTime`, and `handleChangePasscode` are recreated on every render. To prevent unnecessary re-renders of the child list items/switches, we can wrap them in `useCallback`:
```typescript
const toggleBiometry = useCallback(() => {
	logEvent(events.SLC_TOGGLE_BIOMETRY);
	const nextBiometry = !biometry;
	setBiometry(nextBiometry);
	userPreferences.setBool(BIOMETRY_ENABLED_KEY, nextBiometry);
}, [biometry]);

const changeAutoLockTime = useCallback((nextAutoLockTime: number) => {
	logEvent(events.SLC_CHANGE_AUTOLOCK_TIME);
	setAutoLockTime(nextAutoLockTime);
	save(autoLock, nextAutoLockTime);
}, [autoLock, save]);
```

### 2. Protect State Updates in `useEffect` (Unmounted State Guard)
Since `supportedBiometryLabel` and other calls inside the effect are asynchronous, if the component unmounts before they resolve, React could trigger state-update warnings (or cause minor memory leaks). We can guard it using an active flag:
```typescript
useEffect(() => {
	let isMounted = true;
	const init = async () => {
		try {
			const record = await database.servers.get('servers').find(server);
			if (isMounted) {
				serverRecord.current = record;
				setAutoLock(record?.autoLock ?? false);
				setAutoLockTime(
					record?.autoLockTime === null ? DEFAULT_AUTO_LOCK : record?.autoLockTime ?? null
				);
				setBiometry(userPreferences.getBool(BIOMETRY_ENABLED_KEY) ?? DEFAULT_BIOMETRY);
			}
		} catch {
			// noop
		}
		const label = await supportedBiometryLabel();
		if (isMounted) {
			setBiometryLabel(label);
		}
	};
	init();
	return () => {
		isMounted = false;
	};
}, [server]);
```

Otherwise, this is a very clean conversion and the included tests provide great coverage for the state transition edge cases! Keep up the great work! 👍

File: app/views/ScreenLockConfigView.test.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	useTheme: () => ({ theme: 'light', colors: {} })
+}));
+
+jest.mock('../lib/hooks/useAppSelector', () => ({
+	useAppSelector: (selector: (state: any) => unknown) =>
+		selector({
+			server: { server: 'srv' },
+			settings: { Force_Screen_Lock: false, Force_Screen_Lock_After: 0 }
+		})
+}));
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/ScreenLockConfigView.test.tsx:24]: Declining. The suggested `(state: IApplicationState) => unknown` signature doesn't compile as written: the mock passes a partial state (`{ server, settings }`) that isn't assignable to the full `IApplicationState`, so it would need an `as IApplicationState` cast on the literal — which defeats the same type-safety this aims to add. `any` in a test-only selector mock is acceptable here.

File: app/views/ScreenLockConfigView.tsx [hunk 5]
```diff
@@ -28,197 +29,157 @@ interface IItem {
...
+	}, []);
+
+	useEffect(() => {
+		const init = async () => {
+			try {
+				serverRecord.current = await database.servers.get('servers').find(server);
+				setAutoLock(serverRecord.current?.autoLock ?? false);
+				setAutoLockTime(
+					serverRecord.current?.autoLockTime === null ? DEFAULT_AUTO_LOCK : serverRecord.current?.autoLockTime ?? null
+				);
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/ScreenLockConfigView.tsx:58]: Declining for this PR. This is a behavior-preserving class→function migration, and the previous class component initialized `autoLock` the same way — straight from the server record, with no `Force_Screen_Lock` override. Deriving init state from the forced setting would be a behavior change, not a faithful migration. When the lock is enforced, `autoLock` is already persisted to the server record elsewhere. Hardening this should be a separate change.

File: app/views/ScreenLockConfigView.tsx [hunk 6]
```diff
@@ -28,197 +29,157 @@ interface IItem {
...
-		await serversDB.write(async () => {
-			await this.serverRecord?.update(record => {
-				record.autoLock = autoLock;
-				record.autoLockTime = autoLockTime === null ? DEFAULT_AUTO_LOCK : autoLockTime;
+		await database.servers.write(async () => {
+			await serverRecord.current?.update(record => {
+				record.autoLock = nextAutoLock;
+				record.autoLockTime = nextAutoLockTime === null ? DEFAULT_AUTO_LOCK : nextAutoLockTime;
 			});
 		});
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/ScreenLockConfigView.tsx:75]: Declining for this PR. The same optional chaining (`serverRecord?.update(...)`) existed in the class component, so this is pre-existing behavior preserved by the migration, not a regression. In practice the record always loads (you must be connected to a server to reach this screen). A fail-fast guard is reasonable but belongs in a separate, non-behavior-preserving change.

File: app/views/ScreenLockConfigView.tsx [hunk 7]
```diff
@@ -28,197 +29,157 @@ interface IItem {
...
-	renderIcon = () => {
-		const { theme } = this.props;
-		return <List.Icon name='check' color={themes[theme].badgeBackgroundLevel2} />;
+	const handleChangePasscode = async ({ force }: { force: boolean }) => {
+		if (autoLock) {
+			await handleLocalAuthentication(true);
+		}
+		logEvent(events.SLC_CHANGE_PASSCODE);
+		await changePasscode({ force });
 	};
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/ScreenLockConfigView.tsx]: Partially valid, deferring. The floating `save()` in `changeAutoLockTime` is a real runtime-robustness point, but it is not a lint blocker — `no-floating-promises` isn't enabled in this repo (only `no-void`), and CI is green. The handler-wrapping part is lower value: passing an async function to an event prop isn't a floating-promise violation. Leaving as-is for this behavior-preserving refactor; an explicit `.catch` can follow up.

File: app/views/ScreenLockConfigView.tsx [hunk 8]
```diff
@@ -28,197 +29,155 @@ interface IItem {
...
-			biometryLabel: null
-		};
-		this.init();
-	}
-
-	componentWillUnmount() {
-		if (this.observable && this.observable.unsubscribe) {
-			this.observable.unsubscribe();
-		}
+const defaultAutoLockOptions: IItem[] = [
```
Comment (diegolmello) [inline app/views/ScreenLockConfigView.tsx]: This is going to be instantiated on file import and it's going to use the language at that time. After a language change, the text is going to be outdated.

Btw can't we use i18n params? 
```
I18n.t('Local_authentication_auto_lock', { value: 60 })
```

**Review by Shevilll — COMMENTED** [review]
### Refactor Feedback: ScreenLockConfigView Migration to Function Component

Hi @diegolmello, thank you for this excellent migration to a functional component and React hooks! It's highly readable and properly solves the classic hooks state-persistence challenges. 

I've reviewed the diff and have a couple of suggestions that can help prevent potential runtime bugs and optimize performance.

---

#### 1. ⚠️ Potential Safety / Robustness Issue: Safe Destructuring in `handleChangePasscode`
In `app/views/ScreenLockConfigView.tsx`:
```typescript
const handleChangePasscode = async ({ force }: { force: boolean }) => {
    if (autoLock) {
        await handleLocalAuthentication(true);
    }
    logEvent(events.SLC_CHANGE_PASSCODE);
    await changePasscode({ force });
};
```
And it is bound as an `onPress` callback:
```typescript
<List.Item title='Local_authentication_change_passcode' onPress={handleChangePasscode} showActionIndicator />
```

**The Concern:**
In React Native, `onPress` callback handlers typically receive a `GestureResponderEvent` as their first argument. While destructuring `{ force }` from a `GestureResponderEvent` won't crash the application directly (since the event is an object, though `force` will evaluate to `undefined` or touch pressure on older iOS devices rather than a boolean), it can easily cause a crash in tests or custom setups if the `onPress` is called without arguments (e.g., `onPress()`). 

If called with no arguments, the first argument is `undefined`, causing a runtime error:
`TypeError: Cannot destructure property 'force' of 'undefined' as it is undefined.`

**Recommendation:**
Provide a default parameter or handle the optional properties safely to ensure it is resilient under any callback trigger style:
```typescript
const handleChangePasscode = async (args?: { force: boolean }) => {
    const force = args?.force ?? false;
    if (autoLock) {
        await handleLocalAuthentication(true);
    }
    logEvent(events.SLC_CHANGE_PASSCODE);
    await changePasscode({ force });
};
```

---

#### 2. ⚡ Performance Optimization: Memoizing Callbacks & Handlers
Since this component contains multiple interaction states and renders list sections, several internal handlers and elements are recreated on every render:
- `toggleAutoLock`
- `toggleBiometry`
- `handleChangePasscode`
- `changeAutoLockTime`
- `renderItem`

**The Concern:**
Every render creates new references for these functions, which are passed down as props to child components like `Switch`, `List.Item`, etc. This will cause those child components to re-render unnecessarily on any state change (such as toggling auto-lock or biometrics).

**Recommendation:**
Consider wrapping key event handlers with `useCallback` to ensure stable references, and leverage `useMemo` for static items like `items` to keep the UI smooth:
```typescript
const handleChangePasscode = useCallback(async (args?: { force: boolean }) => {
    const force = args?.force ?? false;
    if (autoLock) {
        await handleLocalAuthentication(true);
    }
    logEvent(events.SLC_CHANGE_PASSCODE);
    await changePasscode({ force });
}, [autoLock]);

const toggleAutoLock = useCallback(async () => {
    logEvent(events.SLC_TOGGLE_AUTOLOCK);
    const nextAutoLock = !autoLock;
    setAutoLock(nextAutoLock);
    setAutoLockTime(DEFAULT_AUTO_LOCK);
    if (nextAutoLock) {
        try {
            await checkHasPasscode({ force: false });
            setBiometry(userPreferences.getBool(BIOMETRY_ENABLED_KEY) ?? DEFAULT_BIOMETRY);
        } catch {
            setAutoLock(false);
            setAutoLockTime(DEFAULT_AUTO_LOCK);
            await save(false, DEFAULT_AUTO_LOCK);
            return;
        }
    }
    await save(nextAutoLock, DEFAULT_AUTO_LOCK);
}, [autoLock, save]);
```

Thank you again for your hard work on cleaning up and migrating these views to functional components! Let me know what you think.
